datafusion/physical-expr/src/array_expressions.rs

```diff
@@ -85,7 +87,18 @@ fn array_array(args: &[ArrayRef]) -> Result<ArrayRef> {
         ));
     }
 
-    let data_type = args[0].data_type();
+    let data_type = args
```

Comment:
I think we are trying to avoid casting in the physical-exprs and instead want to cast in the `analyze` pass (so the rest of the optimizer passes see the correct types)

I think this is the relevant code: 

https://github.com/apache/arrow-datafusion/blob/071a2a6dfd43b3b225fce1af8838fa95c79f9ede/datafusion/optimizer/src/analyzer/type_coercion.rs#L389-L397

Perhaps it needs to be taught about array arguments 🤔